examples/petsc/ex28p.cpp

```diff
@@ -0,0 +1,360 @@
+
+   Array<int> block_trueOffsets(3); // number of variables + 1
+   block_trueOffsets[0] = 0;
+   block_trueOffsets[1] = N_space->TrueVSize();
+   block_trueOffsets[2] = L_space->TrueVSize();
+   block_trueOffsets.PartialSum();
+
+   // 8. Define the coefficients of the PDE.
+   ConstantCoefficient u_r_func(1.0);
+   Vector e_r(2);
+   double k0 = M_PI*2/1.0;
+   // We lump the sign, the wavenumber and the dielectric constant into one
+   // coefficient. The dielectric contsant is the square of the refractive
+   // index.
+   e_r(0) = -pow(k0*1.0,2);
+   // This is an example to use different refractive indices in mesh domains
+   e_r(1) = -pow(k0*2.0,2);
+   PWConstCoefficient e_r_func(e_r);
+
+   // 9. Define the parallel grid function and parallel linear forms.
+   BlockVector x(block_offsets);
+   BlockVector trueX(block_trueOffsets);
+
+   // Define the boundary attributes
+   Array<int> ess_bdr;
+   if (pmesh->bdr_attributes.Size())
+   {
+      ess_bdr.SetSize(pmesh->bdr_attributes.Max());
+      ess_bdr = 0;
+   }
+
+   // 10. Assemble the finite element matrices for the LHS and RHS
+   //
+   //                            A = [ Att  0 ]
+   //                                [ 0    0 ]
+   //
+   //                            B = [ Btt Bzt ]
+   //                                [ Btz Bzz ]
+   ParBilinearForm *att = new ParBilinearForm(N_space);
+   ParBilinearForm *btt = new ParBilinearForm(N_space);
+   ParBilinearForm *azz = new ParBilinearForm(L_space);
+   ParBilinearForm *bzz = new ParBilinearForm(L_space);
+   ParMixedBilinearForm *btz = new ParMixedBilinearForm(N_space, L_space);
+
+   PetscParMatrix *pAtt = NULL, *pBtt = NULL, *pBzz = NULL;
+   PetscParMatrix *pAzz = NULL, *pBtz = NULL, *pBzt = NULL;
+   Operator::Type tid = Operator::PETSC_MATAIJ;
+   OperatorHandle Atth(tid), Btth(tid), Bzzh(tid), Btzh(tid), Azzh(tid);
+
+   att->AddDomainIntegrator(new CurlCurlIntegrator(u_r_func));
+   att->AddDomainIntegrator(new VectorFEMassIntegrator(e_r_func));
+   att->Assemble();
+   att->EliminateEssentialBCDiag(ess_bdr, 1.0);
+   att->Finalize();
+   att->ParallelAssemble(Atth);
+   Atth.Get(pAtt);
+   Atth.SetOperatorOwner(false);
+
+   // A dummy Azz is required to set the block size and apply the essential
+   // boundary condition
+   azz->Assemble();
+   azz->EliminateEssentialBCDiag(ess_bdr, 1.0);
+   azz->Finalize();
+   azz->ParallelAssemble(Azzh);
+   Azzh.Get(pAzz);
+   Azzh.SetOperatorOwner(false);
+
+   btt->AddDomainIntegrator(new VectorFEMassIntegrator(u_r_func));
+   btt->Assemble();
+   btt->EliminateEssentialBCDiag(ess_bdr, numeric_limits<double>::min());
+   btt->Finalize();
+   btt->ParallelAssemble(Btth);
+   Btth.Get(pBtt);
+   Btth.SetOperatorOwner(false);
+   (*pBtt) *= -1;
+
+   bzz->AddDomainIntegrator(new DiffusionIntegrator(u_r_func));
+   bzz->AddDomainIntegrator(new MassIntegrator(e_r_func));
+   bzz->Assemble();
+   btt->EliminateEssentialBCDiag(ess_bdr, numeric_limits<double>::min());
```

Comment:
I think this is a duplication of line 221

Reply:
Good catch. Somehow the results were appearing to be correct in my test cases.

Reply:
The `ess_bdr` array is set to zero so the essential boundary conditions are not being applied in any case.

Reply:
That is also a typo. Now when run the example correctly with `ess_bdr = 1;`, I get a zero pivot error when running on one process, but it works fine in parallel. Any clue about what is going on?

Reply:
No if I do not see a stack trace from the PETSc error. Using a very small number on the diagonal is standard practice in eigenvalue problems?

Reply:
Using 0 also works in combination with `skip_zeros=0` when calling `Finalize`. I wanted to do something similar to `ex11p`.

Reply:
This is what I get when I run with `mpirun -np 1`:
```
[0]PETSC ERROR: Zero pivot in LU factorization: https://www.mcs.anl.gov/petsc/documentation/faq.html#zeropivot
[0]PETSC ERROR: Zero pivot row 0 value 2.22507e-305 tolerance 2.22045e-14
```
It does not happen with `mpirun -np 2`

Reply:
Thanks for the explanation. I will try `skip_zeros = 0` and see if it fixes the problem.

Reply:
You are right that the serial and parallel LU solvers have different zero pivot checks. I got it to work in serial by using `-st_pc_factor_shift_type NONZERO`. `skip_zeros = 0` did not have any effect in this case.